perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_tcsMails_mu.java

```diff
@@ -0,0 +1,162 @@
+import cz.metacentrum.perun.core.api.exceptions.WrongReferenceAttributeValueException;
+import cz.metacentrum.perun.core.impl.PerunSessionImpl;
+import cz.metacentrum.perun.core.implApi.modules.attributes.SkipValueCheckDuringDependencyCheck;
+import cz.metacentrum.perun.core.implApi.modules.attributes.UserVirtualAttributesModuleAbstract;
+import cz.metacentrum.perun.core.implApi.modules.attributes.UserVirtualAttributesModuleImplApi;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+/**
+ * Get all emails from Perun for purpose of TCS.
+ *
+ * @author Michal Stava stavamichal@gmail.com
+ */
+@SkipValueCheckDuringDependencyCheck
+public class urn_perun_user_attribute_def_virt_tcsMails_mu extends UserVirtualAttributesModuleAbstract implements UserVirtualAttributesModuleImplApi {
+
+	private static final String preferredMailFriendlyName = "preferredMail";
+	private static final String isMailFriendlyName = "ISMail";
+	private static final String publicMailsFriendlyName = "publicAliasMails";
+	private static final String privateMailsFriendlyName = "privateAliasMails";
+	private static final String o365MailsFriendlyName = "o365EmailAddresses:mu";
+
+	private static final String A_U_D_preferredMail = AttributesManager.NS_USER_ATTR_DEF + ":" + preferredMailFriendlyName;
+	private static final String A_U_D_ISMail = AttributesManager.NS_USER_ATTR_DEF + ":" + isMailFriendlyName;
+	private static final String A_U_D_publicAliasMails = AttributesManager.NS_USER_ATTR_DEF + ":" + publicMailsFriendlyName;
+	private static final String A_U_D_privateAliasMails = AttributesManager.NS_USER_ATTR_DEF + ":" + privateMailsFriendlyName;
+	private static final String A_U_D_o365EmailAddressesMU = AttributesManager.NS_USER_ATTR_DEF + ":" + o365MailsFriendlyName;
+
+	private final static Logger log = LoggerFactory.getLogger(urn_perun_user_attribute_def_virt_tcsMails_mu.class);
+
+	@Override
+	public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDefinition attributeDefinition) throws InternalErrorException {
+
+		Attribute attribute = new Attribute(attributeDefinition);
+		SortedSet<String> tcsMailsValue = new TreeSet<>();
+
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_preferredMail));
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_ISMail));
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_o365EmailAddressesMU));
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_publicAliasMails));
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_privateAliasMails));
+
+		attribute.setValue(new ArrayList(tcsMailsValue));
+
+		return attribute;
+	}
+
+	@Override
+	public List<AuditEvent> resolveVirtualAttributeValueChange(PerunSessionImpl perunSession, AuditEvent message) throws InternalErrorException, WrongReferenceAttributeValueException, AttributeNotExistsException, WrongAttributeAssignmentException {
```

Comment:
Exceptions WrongReferenceAttributeValueException and WrongAttributeAssignmentException are not thrown in this method.

Reply:
Removed.